pkg/providers/instance/vmextension.go

```diff
@@ -0,0 +1,115 @@
+package instance
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute"
+	"github.com/samber/lo"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	vmExtensionType = "Microsoft.Compute/virtualMachines/extensions"
+)
+
+const (
+	cseNameWindows      = "windows-cse-agent-karpenter"
+	cseTypeWindows      = "CustomScriptExtension"
+	csePublisherWindows = "Microsoft.Compute"
+	cseVersionWindows   = "1.10"
+	cseNameLinux        = "cse-agent-karpenter"
+	cseTypeLinux        = "CustomScript"
+	csePublisherLinux   = "Microsoft.Azure.Extensions"
+	cseVersionLinux     = "2.0"
+)
+
+const (
+	aksIdentifyingExtensionName      = "computeAksLinuxBilling"
+	aksIdentifyingExtensionPublisher = "Microsoft.AKS"
+	aksIdentifyingExtensionTypeLinux = "Compute.AKS.Linux.Billing"
+)
+
+// createAKSIdentifyingExtension attaches a VM extension to identify that this VM participates in an AKS cluster
+func (p *DefaultProvider) createAKSIdentifyingExtension(ctx context.Context, vmName string) (err error) {
+	vmExt := p.getAKSIdentifyingExtension()
+	vmExtName := *vmExt.Name
+	log.FromContext(ctx).V(1).Info(fmt.Sprintf("Creating virtual machine AKS identifying extension for %s", vmName))
+	v, err := createVirtualMachineExtension(ctx, p.azClient.virtualMachinesExtensionClient, p.resourceGroup, vmName, vmExtName, *vmExt)
+	if err != nil {
+		log.FromContext(ctx).Error(err, fmt.Sprintf("Creating VM AKS identifying extension for VM %q failed", vmName))
+		return fmt.Errorf("creating VM AKS identifying extension for VM %q, %w failed", vmName, err)
+	}
+	log.FromContext(ctx).V(1).Info(fmt.Sprintf("Created  virtual machine AKS identifying extension for %s, with an id of %s", vmName, *v.ID))
+	return nil
+}
+
+func (p *DefaultProvider) createCSExtension(ctx context.Context, vmName string, cse string, isWindows bool) (err error) {
+	vmExt := p.getCSExtension(cse, isWindows)
+	vmExtName := *vmExt.Name
+	log.FromContext(ctx).V(1).Info(fmt.Sprintf("Creating virtual machine CSE for %s", vmName))
+	v, err := createVirtualMachineExtension(ctx, p.azClient.virtualMachinesExtensionClient, p.resourceGroup, vmName, vmExtName, *vmExt)
+	if err != nil {
+		log.FromContext(ctx).Error(err, fmt.Sprintf("Creating VM CSE for VM %q failed", vmName))
+		return fmt.Errorf("creating VM CSE for VM %q, %w failed", vmName, err)
+	}
+	log.FromContext(ctx).V(1).Info(fmt.Sprintf("Created virtual machine CSE for %s, with an id of %s", vmName, *v.ID))
+	return nil
+}
+
+func (p *DefaultProvider) requiredVMExtensionsInstalled(vm armcompute.VirtualMachine) bool {
+	if len(vm.Resources) < 2 {
+		return false
+	}
+
+	var foundCSE bool // zero values are bool
```

Comment:
This comment doesn't really make sense. It might be cleaner for you to just do:
```
foundCSE := false
foundIdentitfying := false
```

it saves the need for what I think the comment is trying to express.